cibyl/utils/sorting.py

```diff
@@ -0,0 +1,127 @@
+        :type left: Any
+        :param right: Second object to be compared.
+        :type right: Any
+        :return: A negative integer, zero or a positive integer as the left
+            argument is less than, equal to or greater than the right one.
+        :rtype: int
+        """
+        raise NotImplementedError
+
+
+class NativeComparator(Comparator):
+    """Implements a comparison function using Python's native resources.
+    This is, comparison of two objects will be performed through their
+    '__eq__' and '__lt__' methods."""
+
+    @overrides
+    def compare(self, left, right):
+        if left == right:
+            return 0
+
+        return -1 if left < right else 1
+
+
+class SortingAlgorithm(ABC):
+    """Base class for a sorting algorithms. These algorithms are used to
+    reorder the elements of a collection following a set of certain rules.
+    These rules are defines by the comparator in use.
+    """
+
+    def __init__(self, comparator=NativeComparator()):
+        """Constructor.
+
+        :param comparator: Comparison function this uses to determine order.
+        :type comparator: :class:`Comparator`
+        """
+        self._comparator = comparator
+
+    @abstractmethod
+    def sort(self, iterable):
+        """Sorts the given collection in ascending order. The original
+        collection remains untouched, instead, a sorted copy of it is returned.
+
+        :param iterable: The collection to be sorted.
+        :type iterable: :class:`typing.Iterable`
+        :return: The same collection, sorted following this algorithm's
+            comparator.
+        :rtype: list
+        """
+        raise NotImplementedError
+
+
+class BubbleSortAlgorithm(SortingAlgorithm):
+    """Implementation of the Bubble-Sort algorithm. See more at:
+    https://en.wikipedia.org/wiki/Bubble_sort
+    """
+
+    @overrides
+    def sort(self, iterable):
+        def compare(lft, rght):
+            return self._comparator.compare(lft, rght)
+
+        result = list(iterable)  # Do not affect the original list
+
+        for i in range(len(result)):
+            for j in range(0, len(result) - i - 1):
+                if compare(result[j], result[j + 1]) > 0:
+                    temp = result[j]
+
+                    # Swap the two cells
+                    result[j] = result[j + 1]
+                    result[j + 1] = temp
+
+        return result
+
+
+def sort(iterable, algorithm=BubbleSortAlgorithm()):
+    """Shortcut for sorting the elements of a collection.
+
+    :param iterable: The collection to sort.
+    :type iterable: :class:`typing.Iterable`
+    :param algorithm: The algorithm that will be used to sort the collection.
+        Contains the comparator used to determine order.
+    :type algorithm: :class:`SortingAlgorithm`
+    :return: A copy of the input collection, sorted by the algorithm.
+    :rtype: list
+    """
+    return algorithm.sort(iterable)
```

Comment:
That was my first choice, but I could not figure out I way to pass custom comparators to that. That function only supports native Python comparison, with '\__eq__' and '\__lt__'.

On any case, the shorting algorithm that I have implemented is a simple one. I have also added tests checking that it works as it should.

Reply:
Well, there are a couple of reason for it. First, I saw the 'sorted' function, but I did not knew about this one you have linked me to. So I did not see a way of passing a custom comparator to the sorter. Secondly, I was not sure how much performance would be important on this feature. Some queries can run up to thousands of results, so I wanted to have a little more control. Now that I have tried, Bubble-Sort, which is a very simple algorithm, is doing things pretty fast. So thus far I have not had a problem with performance.

On any case, I have tested the sorting algorithm and it works OK. It is a very simple, well-known one, so I do not think it will be much of a problem.

Reply:
You're right, I'll +1 the change since as you said it's well tested and I don't think we'll reach a scale when the performance is a problem. If in the future we were to find ourselves bottlenecked by this (which I highly doubt), here is how the sorted function could be used, for reference: 

```python
from functools import cmp_to_key

from cibyl.models.ci.base.job import Job
from cibyl.outputs.cli.ci.system.sorting.jobs import SortJobsByName

a = SortJobsByName()
jobs = [Job("zjobs"), Job("cjobs"), Job("aajobs")]
sorted_jobs = sorted(jobs, key=cmp_to_key(a.compare))
``